[WIP] Update to 16.60 and add patches for kernel 4.9

brainpower opened the pull request:
What works:
* Building the packages
* Installing the packages
* DKMS install

Workarounds needed:
* Downgrade X to 1.18.4, as the supplied X module is for 1.18

What does not work:
* X crashes with a segfault for me, could be an issue created by the downgrade though:
<pre>
[    50.786] (II) Module "dri2" already built-in
[    50.812] (EE)
[    50.812] (EE) Backtrace:
[    50.813] (EE) 0: /usr/lib/xorg-server/Xorg (OsLookupColor+0x139) [0x59cd49]
[    50.813] (EE) 1: /usr/lib/libc.so.6 (__restore_rt+0x0) [0x7fc8a403a0af]
[    50.813] (EE) 2: ? (?+0x0) [0x0]
[    50.813] (EE)
[    50.813] (EE) Segmentation fault at address 0x0
[    50.813] (EE)
Fatal server error:
[    50.813] (EE) Caught signal 11 (Segmentation fault). Server aborting
</pre>
* VTs are in low resolution (maybe there's a needed patch missing)

Other changes:
* Changed the packages generated to the similar structure the mesa and nvidia packages have, so now there are amdgpu-pro and amdgpu-pro-{dkms,libdrm,libgl,opencl,vdpau,vulkan} and their lib32 counterparts

corngood commented:
Thanks for this.  Is there definitely only one version of the xorg driver shipped with 16.60?  I know in the past they shipped multiple versions, but 1.19 is probably still too new.

Could you run the pre-commit hook from the repo and push again.  It's failing CI because the generated PKGBUILD doesn't match the one in your branch.

Sounds like kms isn't working if your VTs are low res.  Have you verified (e.g. with `lsmod` / `modinfo`) that the dkms module is actually loading?  Anything interesting in dmesg?  You could try kernel options `drm.debug=0x3f` (I think), and `amdgpu.dal=1` (to re-enable dal).

To help debug the crash in Xorg, `strace` and `LD_DEBUG=files` could be useful.

brainpower commented:
> Is there definitely only one version of the xorg driver shipped with 16.60? I know in the past they shipped multiple versions, but 1.19 is probably still too new.
Yeah, they shipped multiple ones in the past, but in 16.60 there is only the one for 1.18.

> Could you run the pre-commit hook from the repo and push again
Sorry, I forgot. I hope this is fixed now.

> Have you verified (e.g. with lsmod / modinfo) that the dkms module is actually loading? 
not with lsmod or modinfo, but X said: 
<pre>[    50.786] (II) AMDGPU(0): AMDGPU kernel module hybrid version: 16.60.3</pre>
Which it probably shouldn't if it couldn't load the module?

It was late in the night at that point, so I didn't investigate long. I'll try the suggested options and commands as soon as I've got time to spare.

brainpower commented:
> > Could you run the pre-commit hook from the repo and push again
> Sorry, I forgot. I hope this is fixed now.

looks like the problem here is, my changes to the gen-PKGBUILD.py made it non-deterministic,
the order of the dependencies was random, because I used set() to remove duplicates... fixed that by sorting the set.

fjfalcon commented:
After makepkg -i:

==> WARNING: The package group has already been built, installing existing packages...
==> Installing amdgpu-pro-installer package group with pacman -U...
loading packages...
resolving dependencies...
warning: cannot resolve "lib32-libomxil-bellagio", a dependency of "lib32-amdgpu-pro"
warning: cannot resolve "lib32-amdgpu-pro=16.60.379184-0", a dependency of "amdgpu-pro"
warning: cannot resolve "lib32-libomxil-bellagio", a dependency of "lib32-amdgpu-pro"
warning: cannot resolve "lib32-amdgpu-pro=16.60.379184-0", a dependency of "amdgpu-pro"
warning: cannot resolve "amdgpu-pro", a dependency of "amdgpu-pro-libgl"
warning: cannot resolve "lib32-libomxil-bellagio", a dependency of "lib32-amdgpu-pro"
warning: cannot resolve "lib32-libomxil-bellagio", a dependency of "lib32-amdgpu-pro"
warning: cannot resolve "lib32-amdgpu-pro", a dependency of "lib32-amdgpu-pro-libgl"
warning: cannot resolve "lib32-libomxil-bellagio", a dependency of "lib32-amdgpu-pro"
warning: cannot resolve "lib32-amdgpu-pro=16.60.379184-0", a dependency of "amdgpu-pro"
warning: cannot resolve "amdgpu-pro=16.60.379184-0", a dependency of "xf86-video-amdgpu-pro"
:: The following packages cannot be upgraded due to unresolvable dependencies:
      amdgpu-pro  amdgpu-pro-libgl  lib32-amdgpu-pro  lib32-amdgpu-pro-libgl  xf86-video-amdgpu-pro

:: Do you want to skip the above packages for this upgrade? [y/N] n
error: failed to prepare transaction (could not satisfy dependencies)
:: lib32-amdgpu-pro: requires lib32-libomxil-bellagio
:: amdgpu-pro: requires lib32-amdgpu-pro=16.60.379184-0
:: lib32-amdgpu-pro: requires lib32-libomxil-bellagio
:: amdgpu-pro: requires lib32-amdgpu-pro=16.60.379184-0
:: amdgpu-pro-libgl: requires amdgpu-pro
:: lib32-amdgpu-pro: requires lib32-libomxil-bellagio
:: lib32-amdgpu-pro: requires lib32-libomxil-bellagio
:: lib32-amdgpu-pro-libgl: requires lib32-amdgpu-pro
:: lib32-amdgpu-pro: requires lib32-libomxil-bellagio
:: amdgpu-pro: requires lib32-amdgpu-pro=16.60.379184-0
:: xf86-video-amdgpu-pro: requires amdgpu-pro=16.60.379184-0
==> WARNING: Failed to install built package(s).

brainpower commented:
You need to install `lib32-libomxil-bellagio` from AUR.

fjfalcon commented:
Yea, with little patch to lib32-libomxil-bellagio config.h i got it installed.
So, got same segfault as you. 

`[    30.227] (II) FBDEV: driver for framebuffer: fbdev
[    30.227] (II) VESA: driver for VESA chipsets: vesa
[    30.228] (II) [KMS] Kernel modesetting enabled.
[    30.228] (WW) Falling back to old probe method for fbdev
[    30.228] (II) Loading sub module "fbdevhw"
[    30.228] (II) LoadModule: "fbdevhw"
[    30.228] (II) Loading /usr/lib/xorg/modules/libfbdevhw.so
[    30.228] (II) Module fbdevhw: vendor="X.Org Foundation"
[    30.228]    compiled for 1.19.1, module version = 0.0.2
[    30.228]    ABI class: X.Org Video Driver, version 23.0
[    30.228] (WW) Falling back to old probe method for vesa
[    30.228] (II) RADEON(0): Creating default Display subsection in Screen section
        "Default Screen Section" for depth/fbbpp 24/32
[    30.228] (==) RADEON(0): Depth 24, (--) framebuffer bpp 32
[    30.228] (II) RADEON(0): Pixel depth = 24 bits stored in 4 bytes (32 bpp pixmaps)
[    30.228] (==) RADEON(0): Default visual is TrueColor
[    30.228] (==) RADEON(0): RGB weight 888
[    30.228] (II) RADEON(0): Using 8 bits per RGB (8 bit DAC)
[    30.228] (--) RADEON(0): Chipset: "HAWAII" (ChipID = 0x67b0)
[    30.228] (II) Loading sub module "fb"
[    30.228] (II) LoadModule: "fb"
[    30.229] (II) Loading /usr/lib/xorg/modules/libfb.so
[    30.229] (II) Module fb: vendor="X.Org Foundation"
[    30.229]    compiled for 1.19.1, module version = 1.0.0
[    30.229]    ABI class: X.Org ANSI C Emulation, version 0.4
[    30.229] (II) RADEON(0): Kernel too old missing accel information, assuming accel is working
[    30.229] (II) Loading sub module "dri2"
[    30.229] (II) LoadModule: "dri2"
[    30.229] (II) Module "dri2" already built-in
[    30.229] (II) Loading sub module "glamoregl"
[    30.229] (II) LoadModule: "glamoregl"
[    30.229] (II) Loading /opt/amdgpu-pro/lib/xorg/modules/libglamoregl.so
[    30.233] (II) Module glamoregl: vendor="X.Org Foundation"
[    30.233]    compiled for 1.18.4, module version = 1.0.0
[    30.233]    ABI class: X.Org ANSI C Emulation, version 0.4
[    30.234] (II) glamor: OpenGL accelerated X.org driver based.
[    30.282] (EE) 
[    30.282] (EE) Backtrace:
[    30.283] (EE) 0: /usr/lib/xorg-server/Xorg (OsLookupColor+0x139) [0x59b7e9]
[    30.283] (EE) 1: /usr/lib/libpthread.so.0 (__restore_rt+0x0) [0x7fd6bb8dc07f]
[    30.284] (EE) 2: ? (?+0x0) [0x0]
[    30.284] (EE) 
[    30.284] (EE) Segmentation fault at address 0x0
[    30.284] (EE) 
Fatal server error:
[    30.284] (EE) Caught signal 11 (Segmentation fault). Server aborting
[    30.284] (EE) 
[    30.284] (EE) 
Please consult the The X.Org Foundation support 
         at http://wiki.x.org
 for help. 
[    30.284] (EE) Please also check the log file at "/var/log/Xorg.0.log" for additional information.
[    30.284] (EE) 
[    30.290] (EE) Server terminated with error (1). Closing log file.
`

Maybe problem in that libfb.so compiled for 1.19.1 version?

brainpower commented:
but: 
> pacman -Qo /usr/lib/xorg/modules/libfb.so
> /usr/lib/xorg/modules/libfb.so is owned by xorg-server
So if you've downgraded X, which you have to, because amd's X module only supports 1.18, then libfb.so should be the one compiled for 1.18 ...
It seems you didn't dowgrade X, but when I tried with X 1.19 there were errors saying the X-ABI version of the server does not match the modules X-ABI version...

fjfalcon commented:
Yea, only noticed that i used wrong log file, cause of startx was started from local user.
Now i get same segfault you get...

brainpower commented:
Well, you could try the LD_DEBUG and running strace on X lice corngood suggested maybe those will get us some information.
Right now I've got some other things to do, otherwise I'd try myself...

fjfalcon commented:
With simple strace got something interesting: 

[strace.txt](https://github.com/corngood/archlinux-amdgpu/files/737164/strace.txt)

For example: 

Fatal server error:
(EE) xf86OpenConsole: Switching VT failed
(EE) 
(EE) 

As for LD_DEBUG=all: 

[lddebug.txt](https://github.com/corngood/archlinux-amdgpu/files/737166/lddebug.txt)

corngood commented:
@brainpower 
> Which it probably shouldn't if the kernel couldn't load the module?

Yeah, that hybrid message suggests the correct module is loaded.  It will sort-of function with the upstream amdgpu, but it will print a warning there instead.

I'm just trying to get this running on NixOS, so I'll see if I have the same problem.

corngood commented:
So on NixOS I managed to get into X with working GL, high res tty etc.  I'm currently a bit out of date (linux-4.8.13 and Xorg-1.18.4), but I'll try updating my system next.

Definitely check your `lsmod`.  There are now three kernel modules needed (amdgpu, amdkcl, and amdttm), whereas in previous versions it was just amdgpu.  The first (and only so far) problem I hit was that I wasn't installing all of them, but that's a different process on NixOS (it doesn't use dkms).

If it's not in there, try `modprobe amdgpu`, and if that fails it will print a bunch of info to `dmesg`.

fjfalcon commented:
lsmod shows that amdttm, amdgpu, amdkcl are present.
Damn. Dunno what to do.

corngood commented:
@fjfalcon do you get a high res tty (i.e. correct kms mode switch)?

fjfalcon commented:
@corngood youp, both displays normaly detected with high res tty.

corngood commented:
Well if one of you could pastebin strace and LD_DEBUG logs of Xorg I can take a look.

fjfalcon commented:
http://fjfalcon.ru/strace.txt from simple strace X
http://fjfalcon.ru/lddebug.txt from LD_DEBUG=all

corngood commented:
The second log has 

> amdgpu: amdgpu_bo_cpu_map failed. (-13)

right before it crashes.

When I did the update on NixOS I ended up dropping all of the patches to the kernel module (they all seemed to be dealt with upstream).  I would try building without them.  I was on linux-4.8, but I'll test 4.9 shortly.

fjfalcon commented:
Also, got same error when tryed to run ethminer: 

amdgpu: amdgpu_bo_cpu_map failed. (-13)
radeonsi: Failed to create a context.
[1]    1484 segmentation fault (core dumped)  /home/fjfalcon/cpp-ethereum/build/ethminer/ethminer -SP 1 -S  -O  -G  128

corngood commented:
Could you get the output of `modinfo amdgpu` and also kcl/ttm?

On 28 Jan 2017 7:42 p.m., Pavel Chernyak <notifications@github.com> wrote:

Also, got same error when tryed to run ethminer:

amdgpu: amdgpu_bo_cpu_map failed. (-13)
radeonsi: Failed to create a context.
[1] 1484 segmentation fault (core dumped) /home/fjfalcon/cpp-ethereum/build/ethminer/ethminer -SP 1 -S -O -G 128

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://github.com/corngood/archlinux-amdgpu/pull/28#issuecomment-275882907>, or mute the thread<https://github.com/notifications/unsubscribe-auth/AC7z_iwKidVqO7oK7RtQ5CrmHVvDoQ4Yks5rW9JtgaJpZM4LwZy2>.

fjfalcon commented:
http://fjfalcon.ru/amdgpu.txt
http://fjfalcon.ru/amdkcl.txt
http://fjfalcon.ru/amdttm.txt

corngood commented:
Well it's loading them all from updates/ so that seems fine.  The only things I can think of trying are:

- remove the patches from the dkms module (except for maybe the OS detection and build related ones)
- test it with linux-4.8

On 28 Jan 2017 7:55 p.m., Pavel Chernyak <notifications@github.com> wrote:

http://fjfalcon.ru/amdgpu.txt
http://fjfalcon.ru/amdkcl.txt
http://fjfalcon.ru/amdttm.txt

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://github.com/corngood/archlinux-amdgpu/pull/28#issuecomment-275883489>, or mute the thread<https://github.com/notifications/unsubscribe-auth/AC7z_uJ9tYEZs3WmVFTBRdd_Td3-T8uVks5rW9VfgaJpZM4LwZy2>.

brainpower commented:
Well, amdgpu_bo_cpu_map is defined in amdgpu-pro-libdrm/opt/amdgpu-pro/include/libdrm/amdgpu.h, and it's in libdrm_amdgpu.so:
<pre>
$ strings amdgpu-pro-libdrm/opt/amdgpu-pro/lib/x86_64-linux-gnu/libdrm_amdgpu.so | grep amdgpu_bo_cpu_map
amdgpu_bo_cpu_map
amdgpu_bo_cpu_map
</pre>

According to [1], it probably does not much except calling drmCommandWriteRead, which in turn calls drmIoctl and that does some ioctl which is probably the thing that fails... were there any patches for 4.8 related to ioctl things which might be missing? Does DAL have anything to do with that?
I don't know much about how that works...

[1] https://lists.freedesktop.org/archives/dri-devel/2015-April/081539.html

fjfalcon commented:
Well, with 4.8 kernel same problem - not working. 
http://fjfalcon.ru/strace4.8.txt
http://fjfalcon.ru/lddebug4.8.txt

fjfalcon commented:
Got it working with 4.8. 
Looks like problem was with /usr/lib folder. There was old packages that wasn't replaced.
After i copy libs from /opt/amdgpu-pro/lib/x86_64-linux-gnu xorg started as it should.

Looks like package totally not ready for production =(

corngood commented:
Well sounds like it's pretty close to working if you fixed it.  Remember this is just a PR.

Could you `pacman -Qo` those files you needed to replace so we know which package owned them?  Then it should just be a matter of adding a conflict and possibly symlinking them if they aren't in LD_LIBRARY_PATH.

corngood commented:
The libraries in your LD_DEBUG log that look suspect are:

* /usr/lib/libGL.so.1
* /usr/lib/libgbm.so.1
* /usr/lib/libdrm_amdgpu.so.1
* /usr/lib/libdrm_radeon.so.1

Could you find package owners for those?  They should all be from amdgpu-pro.

fjfalcon commented:
They were from libdrm/lib32-drm

29 янв. 2017 г. 4:55 пользователь "David McFarland" <
notifications@github.com> написал:

> Well sounds like it's pretty close to working if you fixed it. Remember
> this is just a PR.
>
> Could you pacman -Qo those files you needed to replace so we know which
> package owned them? Then it should just be a matter of adding a conflict
> and possibly symlinking them if they aren't in LD_LIBRARY_PATH.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/corngood/archlinux-amdgpu/pull/28#issuecomment-275888271>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AA_E16_lsjcWOU0BZVuU8bX5dm84eCxVks5rW_GIgaJpZM4LwZy2>
> .
>

fjfalcon commented:
Also, opencl libs also was missing. Looks like nothing for copied to line
or preload path

29 янв. 2017 г. 5:00 пользователь "Pavel Chernyak" <fjfalcon@fjfalcon.ru>
написал:

> They were from libdrm/lib32-drm
>
> 29 янв. 2017 г. 4:55 пользователь "David McFarland" <
> notifications@github.com> написал:
>
>> Well sounds like it's pretty close to working if you fixed it. Remember
>> this is just a PR.
>>
>> Could you pacman -Qo those files you needed to replace so we know which
>> package owned them? Then it should just be a matter of adding a conflict
>> and possibly symlinking them if they aren't in LD_LIBRARY_PATH.
>>
>> —
>> You are receiving this because you were mentioned.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/corngood/archlinux-amdgpu/pull/28#issuecomment-275888271>,
>> or mute the thread
>> <https://github.com/notifications/unsubscribe-auth/AA_E16_lsjcWOU0BZVuU8bX5dm84eCxVks5rW_GIgaJpZM4LwZy2>
>> .
>>
>

MartB commented:
I can confirm what @fjfalcon said.
After installing this packages on a totally fresh arch linux install and downgrading the xorg server to 18.4.
The libraries

```
/usr/lib/libgbm.so.1
/usr/lib/libdrm_amdgpu.so.1
/usr/lib/libdrm_radeon.so.1
```
are owned by libdrm2.4.74-1 and after copying the files from the folder he mentioned xorg starts up just fine.

Kind regards
Martin.

corngood commented:
@brainpower Could you add conflicts and provides for libdrm where appropriate and see if it fixes it for you without having to move things around?

@MartB Thanks for verifying that.

Here's a checklist of things that could be tested:

- [ ] OpenGL x64
- [ ] OpenGL x86
- [ ] Vulkan x64
- [ ] Vulkan x86
- [ ] OpenCL x64
- [ ] OpenCL x86
- [ ] VDPAU x64
- [ ] VDPAU x86
- [ ] linux-4.9

So could you guys post if you get positive results with any of those?  Also could you post your GPU model?

Thanks

MartB commented:
I can confirm 4.9.6-1
```
> modinfo amdgpu
filename:       /lib/modules/4.9.6-1-ARCH/updates/amdgpu.ko
version:        16.60.3
license:        GPL and additional rights
description:    AMD GPU
author:         AMD linux driver team
```

corngood commented:
Could you check that these libraries (from your log) are owned by the amdgpu-pro packages?

/usr/lib/libGL.so.1
/usr/lib/libgbm.so.1
/usr/lib/libdrm_amdgpu.so.1
/usr/lib/xorg/modules/dri/radeonsi_dri.so

If not, which package owns them?

On 28 Jan 2017 6:52 p.m., Pavel Chernyak <notifications@github.com> wrote:

http://fjfalcon.ru/strace.txt from simple strace X
http://fjfalcon.ru/lddebug.txt from LD_DEBUG=all

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://github.com/corngood/archlinux-amdgpu/pull/28#issuecomment-275880506>, or mute the thread<https://github.com/notifications/unsubscribe-auth/AC7z_ixJrXNq39eVpD9redaPBrNSTiTrks5rW8aOgaJpZM4LwZy2>.

MartB commented:
```
pacman -Qo /usr/lib/libGL.so.1
/usr/lib/libGL.so.1 amdgpu-pro-libgl 16.60.379184-0

pacman -Qo /usr/lib/libgbm.so.1
/usr/lib/libgbm.so.1 mesa 13.0.3-1 

pacman -Qo /usr/lib/libdrm_amdgpu.so.1
/usr/lib/libdrm_amdgpu.so.1  libdrm 2.4.74-1

pacman -Qo /usr/lib/xorg/modules/dri/radeonsi_dri.so
/usr/lib/xorg/modules/dri/radeonsi_dri.so ist in mesa 13.0.3-1 enthalten
```

pavanky commented:
How did you guys end up downgrading to 1.18 ? I keep getting 
```
cannot resolve "xf86-input-driver", a dependency of "xorg-server"
```

fjfalcon commented:
@pavanky Udd on install with downgrading xf86-input-evdev xf86-input-evdev-2.10.4-1-x86_64.pkg.tar.xz

corngood commented:
@fjfalcon ethminer is OpenCL, right?  It's probably due to the conflict with ocl-icd.  What happens if you use the amdgpu-pro version of libOpenCL.so?

corngood commented:
@fjfalcon So this worked ok for you with linux-4.9?  It worked for me on 4.8, but on 4.9 I'm getting a crash in the dri driver when xorg starts: http://ix.io/1RrB

brainpower commented:
* I've added the conflicts and provides for libdrm,
* I added the /opt/amdgpu-pro/lib/*-linux-pc/ paths to /etc/ld.so.conf.d, so the so's inside will be picked up by the dynamic linker
* I've re-added the patch which disables DAL by default, which got me the high-res tty back

But now I get the same X crash as @corngood .

Here are strace and LD_DEBUG (the last 20000 lines) output:
[lddebug2.txt](https://github.com/corngood/archlinux-amdgpu/files/737958/lddebug2.txt)
[x_strace.txt](https://github.com/corngood/archlinux-amdgpu/files/737953/x_strace.txt)

corngood commented:
@brainpower I see in that log

> /usr/lib/x86_64-linux-gnu/dri/amdgpu_dri.so

Is that correct?

corngood commented:
@brainpower I checked mine and it's definitely loading the correct amdgpu_dri.so, but still crashing, so it's probably something else.

Let me know if just switching to 4.8 fixes it.

brainpower commented:
The path of the dri.so should be correct, `libgl1-amdgpu-pro-dri_16.60-379184_amd64.deb´ installs it into /usr, not /opt ... so yeah, must be sth else...

thevoidnn commented:
hi, thx for the pkgbuild!

HD7870 @ linux 4.10.0-rc5-mainline here
I'm on 0d5eec3ce08aedb80798cd71aa090e850e65b23e (w/o vdpau fix)

builds, installs (won't complie with dkms, so kernel module is from mainline)
Xorg (1.18.4) works, Steam works, DOOM works :)
no vulkan yet: failed with VK_ERROR_OUT_OF_HOST_MEMORY

corngood commented:
@thevoidnn Thanks for testing.  I suggest giving it a try with linux-4.8, unless you need 4.10 for other reasons.  The open source driver usually only _sort of_ works with the amdgpu-pro userspace, and you might get Vulkan then.

apemanzilla commented:
Trying to build with `makepkg -s` and I'm getting a permissions error. Problem with the PKGBUILD?

```
==> Starting package_amdgpu-pro-vdpau()...
ln: failed to create symbolic link '/usr/lib/libvdpau_amdgpu.so.1.0.0': Permission denied
==> ERROR: A failure occurred in package_amdgpu-pro-vdpau().
    Aborting...
```

Astilex5 commented:
I am setting up a new Arch Machine and now that the 4.9 Kernel in Arch has CIK and SI [enabled](https://git.archlinux.org/svntogit/packages.git/commit/?h=packages/linux) by default I can help more with working on it. What needs testing?

corngood commented:
@apemanzilla looks like the last push also failed with that error in CI.  @brainpower do you know what that's about?

@Astilex5 that CIK/SI support will allow you to use the open source amdgpu driver, but it's not relevant for amdgpu-pro.  In fact, none of us have got amdgpu-pro 16.60 working on linux-4.9 yet.  It only appears to be stable on linux-4.8.  So if you want to test amdgpu-pro you'll need to downgrade to linux-4.8 and xorg-1.18.

Astilex5 commented:
@corngood Hmmm so I do not need CIK and SI for AMDGPU-PRO? I am using GCN 1.0 hardware. Will that work on 4.8 with AMDGPU-PRO? (In theory that is?)

corngood commented:
@Astilex5 amdgpu-pro has it's own kernel driver which has supported CIK/SI for quite some time, so yes it should work.  However, it looks like there's a problem with the latest PKGBUILD in this PR.

I suggest keeping an eye on this PR for positive news, and in the meantime install mesa-libgl and xf86-video-amdgpu.  That'll give you the full open source stack, which is pretty solid these days.

apemanzilla commented:
@Astilex5 @corngood I'm not sure if it's changed but I had to enable CIK to get my R9 390 to work with AMDGPU-PRO 16.30.

Anyways, once the PKGBUILD is fixed I'll test it out.

Astilex5 commented:
@corngood Thanks for the solid info I will install the OS drivers for now and keep an eye on this.

corngood commented:
@apemanzilla That's probably because you weren't using the dkms driver.  I think it was disabled by default in the 16.30 PKGBUILD.  The amdgpu-pro userspace will 'do stuff' with the upstream driver.  In my experience it will mostly work, with reduced performance, and severe stability problems in Vulkan.  It was probably 16.30 / linux-4.7 when I last tested that, and at that point there were important system calls missing.

Now the situation may be better with 4.9/10, so if anyone wants to give it a try, just remove the dkms driver and make sure (upstream) amdgpu is being loaded.  If you do that, posting the Xorg log and `glxinfo` would be appreciated.  It would also be a good idea to test some serious stuff like modern games and the Vulkan smoketest.

MartB commented:
I checked out 0d5eec3ce08aedb80798cd71aa090e850e65b23e and it builds and installs just fine on 4.9.6-1.
However, im missing a way to verify that its indeed using the amd gpu pro driver.

Besides modinfo telling me the following:
```
filename:       /lib/modules/4.9.6-1-ARCH/updates/amdgpu.ko
version:        16.60.3
license:        GPL and additional rights
description:    AMD GPU
author:         AMD linux driver team
```
vulkanCapsViewer errors out with No compatible Vulkan driver found.
I have no idea what vulkan API version i need from here: https://vulkan.lunarg.com/sdk/home#linux to get it to run.

corngood commented:
@MartB Yeah that's the right driver (it's in updates/ and has the correct version number).

Which GPU are you using?  I don't think anyone else has actually got into X with 4.9.  Would you mind posting your Xorg log?

I'm using an R9 290, and it doesn't work.

For vulkan, `VK_LOADER_DEBUG=all` can help you trace the loader.  It needs to find the manifest files (should be in `/usr/share/vulkan`) and then the libraries they point to.

MartB commented:
Im using an R9 290X from Asus.
Xorg.0.log
http://pastebin.com/hJHqpVYW

Checking vulkan now.

corngood commented:
From your Xorg log:

```
[    10.917] (II) AMDGPU(0): AMDGPU DDX hybrid version: 16.60.3
[    10.917] (EE) AMDGPU(0): Failed to open amdgpu hybrid version
```

I believe this means that it's using the builtin kernel module and not the dkms one.  I'm not sure why your modinfo indicated otherwise, but it explains why it's not crashing for you.

AustinJ235 commented:
I have tested this with an RX 480. Xorg fails to start with and without the dkms module on the stock 4.9 kernel. Not sure if I am missing anything. This arch install has never been updated to xorg 1.19. Seems that I have a similar segfault as some before, but mine has a long backtrace for some reason.

**with dkms:**
xorg https://austinj.me/logs/xorg_49_stock_with_dkms.txt
ld_debug=all https://austinj.me/logs/ld_debug_49_stock_with_dkms.txt
strace https://austinj.me/logs/strace_49_stock_with_dkms.txt

**without dkms**
xorg https://austinj.me/logs/xorg_49_stock_without_dkms.txt
ld_debug=all https://austinj.me/logs/ld_debug_49_stock_without_dkms.txt
strace https://austinj.me/logs/strace_49_stock_without_dkms.txt

corngood commented:
@AustinJ235 Yeah, that's interesting.  You're getting the same hybrid driver error @MartB, but it wasn't crashing for him.  I suppose it could be gpu dependent, but for it to be more broken on Polaris is... odd.

Has anyone had a chance to verify that switching to linux-4.8 fixes it?  It'll probably be the weekend before I get a chance to test that myself.

MartB commented:
@corngood thanks for pointing out this embarassing mistake :+1: 
@everyone i made a horribly mistake while testing I had the amdgpu module in my mkinitcpio.conf and forgot to remove it :/

I removed it and now it crashes with the same stacktrace @AustinJ235 posted up there.
Sorry for wasting everyones time!

AustinJ235 commented:
4.8.13 seems to have the same issue. My logs from before were actually using the kernel built from agd5f amd-staging-4.9 branch. It seems to return the same results as stock 4.9.

**with dkms**
xorg https://austinj.me/logs/xorg_48_stock_with_dkms.txt
ld_debug https://austinj.me/logs/ld_debug_48_stock_with_dkms.txt
strace https://austinj.me/logs/strace_48_stock_with_dkms.txt

**without dkms**
xorg https://austinj.me/logs/xorg_48_stock_without_dkms.txt
ld_debug https://austinj.me/logs/ld_debug_48_stock_without_dkms.txt
strace https://austinj.me/logs/strace_48_stock_without_dkms.txt

mozolny commented:
Hi, I tested latest working 16.60 version on 3 kernels with Xorg and basic 3D appliciation (Furmark, glxinfo). GPU is RX460.

1. 4.8.14-1-zen
   * **dkms** - works [Xorg log](http://pastebin.com/btaS611b)
   * **builtin** - works
2. Xorg.4.9.6-1-ARCH
   * **dmks** - Xorg crash (Xorg log accidentally deleted)
   * **builtin** - didn't test (I assume it works)
3. Xorg.4.10.0-rc5-mainline
   * **dmks** module doesn't build
   * **builtin** - works [Xorg log](http://pastebin.com/Aa7NVMaf)

Shame on me that the most interesting log file has been lost (during writing comment), I can regenerate it (and make strace log) if somebody interested in another one report. But it looks 4.9 is just not supported.
 
I run into mentioned problem with builtin amdgpu module loading at start, while _modinfo_ was reporiting dkms version. I think it was _mkinitcpio_ issue, but even after booting fallback initrd image and rebuilding all images, the builtin module was loaded automatically and preventing dkms module to load (at 4.8). Only solution was blacklisting amdgpu and modprobing it manually before starting X. However at 4.9 the dkms module is loaded automatically without blacklisting hack.

corngood commented:
@mozolny Thanks for testing all of that.

Does anyone have a strong desire to get this working on 4.9, vs. just requiring 4.8?  We already require Xorg to be rolled back.  Personally I'd be fine with 4.8, since most of what interests me about 4.9/10 are the amdgpu changes.  We could try to fix it for 4.10, but I don't have high hopes of that solving the crash.

We should be able to enforce both of those rollbacks in PKGBUILD conflicts.  xorg-server has `X-ABI-VIDEODRV_VERSION=23` in the current version.  I think the one we want is 20???

fjfalcon commented:
Hi there, all my tasks(xorg, mining, working) works on 4.8.13 and xorg
1.18. So, driver looks stable for this versions...

1 февр. 2017 г. 1:57 AM пользователь "David McFarland" <
notifications@github.com> написал:

> @mozolny <https://github.com/mozolny> Thanks for testing all of that.
>
> Does anyone have a strong desire to get this working on 4.9, vs. just
> requiring 4.8? We already require Xorg to be rolled back. Personally I'd be
> fine with 4.8, since most of what interests me about 4.9/10 are the amdgpu
> changes. We could try to fix it for 4.10, but I don't have high hopes of
> that solving the crash.
>
> We should be able to enforce both of those rollbacks in PKGBUILD
> conflicts. xorg-server has X-ABI-VIDEODRV_VERSION=23 in the current
> version. I think the one we want is 20???
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/corngood/archlinux-amdgpu/pull/28#issuecomment-276520892>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AA_E1zmjGpqeF_zdWWZAVHT6m0dqT9pAks5rX7xzgaJpZM4LwZy2>
> .
>

apemanzilla commented:
If you don't mind my newbiness, how are you guys fixing the permissions error that's causing the Travis build to fail?

MartB commented:
@apemanzilla I guess most of the people here did not.
I just checked out 0d5eec3ce08aedb80798cd71aa090e850e65b23e and built that one instead, thats the easiest solution.

Kind regards
Martin

brainpower commented:
sorry @all for not reacting at all, but I've been seriously sick the last couple of days and I've still not recovered completely, I'll be resting a few more days and be back when I've recovered.

Gelmo commented:
@corngood 
I don't think we should be targeting anything prior to 4.9 considering that 4.9 has SI and CIK enabled by default. This package should be as vanilla as possible; there shouldn't be any need to change the build config (then we could use the dkms module regardless of the kernel we're on, as long as xorg-server supports it). If the plan is to use dkms, this should not be recompiling the kernel. Additionally, we're on Arch for rolling releases; if you're going to explicitly get away from Core, the only reasonable version to target would be LTS, which is currently 4.4 on Arch and will likely be 4.9 in the near future. Targeting 4.8 will be a huge issue for anyone who intends on using new hardware in the future, unless you expect them to write original patches.

Also, I'm new to DKMS and do not have a firm understanding of it yet; Unless we are explicitly blacklisting the amdgpu built-in module, do we explicitly need to NOT have amdgpu in /etc/mkinitcpio.conf or should we have it there? I'm just confused by this comment:

https://github.com/corngood/archlinux-amdgpu/pull/28#issuecomment-276486648

corngood commented:
@Gelmo The SI/CIK support in 4.9 doesn't matter here, because the dkms driver replaces the one provided with the kernel.  The dkms driver supports SI/CIK, and has done for the last few versions.

As for 4.4, if it works I don't see why we couldn't set the constraint to whatever range is supported with the driver.  So far it seems < 4.9 is required, but I don't know what the lower bound is.

> Unless we are explicitly blacklisting the amdgpu built-in module, do we explicitly need to NOT have amdgpu in /etc/mkinitcpio.conf or should we have it there? 

I can't answer this confidently without testing.  Obviously the safe thing to do is leave it out, but if mkinitcpio pulls in the module from updates/ the same way that modprobe does, it should work.

Gelmo commented:
@corngood 
Please excuse my misunderstanding; that definitely makes sense (in regard to the dkms driver supporting SI/CIK regardless of kernel config).

Regarding the mkinitcpio options, I'll have to try it out when we get it working on 4.9 and report the results. Thanks for the quick response.

corngood commented:
@Gelmo  I tested this with linux-4.4, and the module builds, but I get:

```
amdkcl: exports duplicate symbol drm_atomic_helper_duplicate_state (owned by drm_kms_helper)
```

This was on NixOS though, so you want to try it out with the LTS kernel.

That error in itself doesn't look like a hard thing to solve, but it's not a good sign, and I don't really want to put a lot of time into getting this working if it's not meant for 4.4.

brainpower commented:
I pushed the fix for the permission problem
and added the conflicts which should enforce X 1.18.

I had also tested with with linux-4.8 before I got sick, and I remember X starting, but crashing later as soon as I tried to start Steam.
Well, I wouldn't like to also require an old version of the kernel, but if there's no other solution we may have to...
On the other side, I found a patch which claims to get 16.50 to work on 4.9 [1], which I want to look at again before giving up on 4.9.

[1] https://gist.github.com/pimlie/064475bbd858995feb38d620d17e2437

corngood commented:
Thanks @brainpower 

If anyone has a chance to test this latest PKGBUILD with steam (or other 32-bit GL apps), let us know how it goes.

mozolny commented:
I can't test today latest version, but remember 2 issues with previous working and Call of Duty / wine:

-missing 32-bit dri driver module (`/usr/lib32/dri/amdgpu(...).so`)
-missing/conflicting _lib32-opencl-icd-loader_ metamodule which perhaps should be provided by _lib32-amdgpu-pro-opencl-icd_

brainpower commented:
@mozolny 
77a34e6 and  6756ab4 should fix those two issues

corngood commented:
@brainpower FYI looks like something is failing in the arch-travis script.  Master doesn't even build at the moment.  I'll try to fix it when I get a chance.

corngood commented:
@brainpower could you cherry-pick ea02acbd302f567335eb97412f12212d86a57ad4 and push again?  That should fix the CI error.

mozolny commented:
I confirm **4.9** is working with **dkms** (2D, 3D) module after 84568240ae7124c2da73a0e37745d974ef1d1425
[Xorg log for curious](http://pastebin.com/9xQ9imYy)

Thank You @brainpower for fixes, but 32-bit `amdgpu_dri.so` is now placed in wrong directory (`/usr/lib32` instead of `/usr/lib32/dri`)

I think `lib32-amdgpu-pro-opencl` should maybe provide both `lib32-opencl-icd-driver` and `lib32-opencl-icd-loader` (as it contains _libOpenCL.so_ implementation). Or perhaps `lib32-amdgpu-pro-opencl` can be installed along of `lib32-ocl-icd`  (as _libOpenCL.so_ is in different path). And there is also `lib32-libcl` metapackage.
I don't know, I don't use OpenCL anyway (especially 32-bit); I needed it only to install `wine-staging-git`.

brainpower commented:
ah, ´mv´ bit me in the ass, not doing what I expected it to do...

@ opencl: I don't know too much about how openCL works myself. I've never consciously used it.

If I understood correctly, the libOpenCL.so (also called icd-loader) is some kind of wrapper, which delegates the things the user wants to the driver, which is the "real" openCL implementation. For that it looks into ´/etc/OpenCL/vendors/´ and selects an appropriate vendor library from there and delegates the real work to that driver.

At the moment it would use both amdgpu-pro's ´libOpenCL.so´ and the driver ´libamdocl64.so´, because I added ´/opt/amdgpu-pro/lib/x86_64-linux-gnu/´ to the paths where the dynamic linker looks for libraries.
But the ´ocl-icd´ should work, too. Because it should look into ´etc/OpenCL/vendors/amdocl64.icd´ to get the path for the driver lib and also delegate all the work to ´libamdocl64.so´.

Well, if there's no difference in that delegating, it shouldn't matter which icd-loader is used.
You could test with LD_PRELOAD=/usr/lib/libOpenCL.so.1 if it makes a difference when using the loader from ´ocl-icd´...

fjfalcon commented:
btw, with ocl-icd from arch repo everything work as it should.

On 7 February 2017 at 02:40, brainpower <notifications@github.com> wrote:

> ah, ´mv´ bit me in the ass, not doing what I expected it to do...
>
> @ opencl: I don't know too much about how openCL works myself. I've never
> consciously used it.
>
> If I understood correctly, the libOpenCL.so (also called icd-loader) is
> some kind of wrapper, which delegates the things the user wants to the
> driver, which is the "real" openCL implementation. For that it looks into
> ´/etc/OpenCL/vendors/´ and selects an appropriate vendor library from there
> and delegates the real work to that driver.
>
> At the moment it would use both amdgpu-pro's ´libOpenCL.so´ and the driver
> ´libamdocl64.so´, because I added ´/opt/amdgpu-pro/lib/x86_64-linux-gnu/´
> to the paths where the dynamic linker looks for libraries.
> But the ´ocl-icd´ should work, too. Because it should look into
> ´etc/OpenCL/vendors/amdocl64.icd´ to get the path for the driver lib and
> also delegate all the work to ´libamdocl64.so´.
>
> Well, if there's no difference in that delegating, it shouldn't matter
> which icd-loader is used.
> You could test with LD_PRELOAD=/usr/lib/libOpenCL.so.1 if it makes a
> difference when using the loader from ´ocl-icd´...
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/corngood/archlinux-amdgpu/pull/28#issuecomment-277850865>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AA_E180tI2B8gPAJYkrn3ELcYWEoPUIqks5rZ69-gaJpZM4LwZy2>
> .
>



-- 
С уважением,
Черняк Павел Александрович

MartB commented:
Can confirm builds, installs and starts on 4.9 like a charm now!

corngood commented:
@brainpower great work on those patches.

This is already a big improvement to what's on AUR, but is there anything you want to clean up before we release it?

Also let me know if you want push rights on the AUR package and/or this git repo (I can either give you permissions or you can make a fork).  It's fine with me either way.

brainpower commented:
re-using corngoods checklist I can confirm the following things working on my machine:

* [x] linux-4.9
* [x] linux-4.9 AND Xorg
* [x] OpenGL x64
* [ ] OpenGL x86
* [ ] Vulkan x64
* [ ] Vulkan x86
* [ ] OpenCL x64
* [ ] OpenCL x86
* [x] VDPAU x64
* [ ] VDPAU x86


vkQuake still somehow used RADV instead of the AMDGPU-PRO vulkan driver:
<pre>
Vulkan Initialization
WARNING: radv is not a conformant vulkan implementation, testing use only.
Vendor: AMD
Device: AMD RADV POLARIS10
</pre>
So there seems to still be some issues...
Maybe ´/etc/vulkan/icd.d/amd_icd64.json´ is the wrong location for the icd file? Should it be /usr/share/vulkan/icd.d ?
I'll test that now.

There are a few things that could be cleaned up before releasing this to AUR, my overwirting PKGEXT in the PKGBUILD and pkgrel being 0 for example... I'll do those soon.

Give me permissions if you want, but I'm also fine with just issuing PRs. ^^
And I don't know if I want the responsibility of being the official AUR maintainer for this...

AustinJ235 commented:
I am still getting the segfault issue with xorg. Is there something wrong with my setup?

https://austinj.me/logs/fd875f1_xorg_log.txt

brainpower commented:
Did you run mkinitcpio after updating?
If not, your kernel is probably loading an old version of the dkms module.

AustinJ235 commented:
I have not done that in the past. Tried that after installing the packages and still has the same issue.

corngood commented:
@brainpower That fix to `amdgpu_verify_access` did the trick for me on NixOS.  I'm now up and running on 4.9.6.  I also let the guy who maintains the gentoo patches know.  I think he's already working on 4.10.  That'll be the next thing :)

Let me know when you want this pushed to AUR.  I'm happy to keep doing that.

AustinJ235 commented:
I have a 270 laying around I'll give that a shot to see if I experience the same issue. Maybe it is just related to the rx 480.

brainpower commented:
@corngood Yeah, I took the first version of that patch from some nvidia thread with a fix for their kernel module but after reading the patch I mentioned [1] I saw that that guy did not just cast the struct, but extracted the private_data member... Which seemed more correct to me... 

I don't have any additions for now and for most people it seems to work,
so I'm okay with merging this and putting this to AUR.
I'd say: Let's use new Issues or PRs for any further issues and close this one, it has grown quite big anyway.

[1] https://gist.github.com/pimlie/064475bbd858995feb38d620d17e2437

corngood commented on gen-PKGBUILD.py:
```diff
@@ -7,28 +7,287 @@
+
+		'xf86-video-amdgpu-pro': Package(
+			desc = "The AMDGPU Pro X.org video driver",
+			conflicts = ['xf86-video-amdgpu', 'xorg-server<1.18.0', 'xorg-server>=1.19.0' 'X-ABI-VIDEODRV_VERSION<20', 'X-ABI-VIDEODRV_VERSION>=21'],
```
@brainpower I just noticed this typo here (missing `,`).  It's not an urgent problem as it looks like it will only break the upper bound of xorg-server and the lower bound of the XBI version.

brainpower replied:
As it's a minor thig, I fixed it in master directly.
